# Jupyter

We'll be using Jupyter for all of our examples&mdash;this allows us to run python in a web-based notebook, keeping a history of input and output, along with text and images.

For Jupyter help, visit:
https://jupyter.readthedocs.io/en/latest/content-quickstart.html

We interact with python by typing into _cells_ in the notebook.  By default, a cell is a _code_ cell, which means that you can enter any valid python code into it and run it.  Another important type of cell is a _markdown_ cell.  This lets you put text, with different formatting (italics, bold, etc) that describes what the notebook is doing.

You can change the cell type via the menu at the top, or using the shortcuts:

  * ctrl-m m : mark down cell
  * ctrl-m y : code cell

Some useful short-cuts:

 * shift+enter = run cell and jump to the next (creating a new cell if there is no other new one)
 * ctrl+enter = run cell-in place
 * alt+enter = run cell and insert a new one below


In [2]:
print("hello")

hello


A "markdown cell" enables you to typeset LaTeX equations right in your notebook.  Just put them in <span>$</span> or <span>$$</span>:

$$e^{ix} = \cos x + i \sin x$$

<div class="alert alert-block alert-warning">
    <b>Important</b> : 
    
When you work through a notebook, everything you did in previous cells is still in memory and </em>known</em> by python, so you can refer to functions and variables that were previously defined.  
    
Even if you go up to the top of a notebook and insert a cell, all the information done earlier in your notebook session is still defined&mdash;it doesn't matter where physically you are in the notebook.  If you want to reset things, you can use the options under the _Kernel_ menu.
</div>

<div class="alert alert-block alert-info"><h3><span class="fa fa-flash"></span> Quick Exercise:</h3>

Create a new cell below this one.  Make sure that it is a _code_ cell, and enter the following code and run it:
```
    
 print("Hello, World")

    
```
</div>

`print()` is a _function_ in python that takes arguments (in the `()`) and outputs to the screen.  You can print multiple quantities at once like:

In [3]:
print(1, 2, 3)

1 2 3


# Basic Datatypes

Now we'll look at some of the basic datatypes in python&mdash;these are analogous to what you will find in most programming languages, including numbers (integers and floating point), and strings.

Some examples come from the python tutorial:
http://docs.python.org/3/tutorial/

## integers

Integers are numbers without a decimal point.  They can be positive or negative.  Most programming languages use a finite-amount of memory to store a single integer, but in python will expand the amount of memory as necessary to store large integers.

The basic operators, `+`, `-`, `*`, and `/` work with integers

In [4]:
2+2+3

7

In [5]:
2*-4

-8

<div class="alert alert-block alert-warning">
    
Note: integer division is one place where python 2 and python 3 different
    
In python 3.x, dividing 2 integers results in a float.  In python 2.x, dividing 2 integers results in an integer.  The latter is consistent with many strongly-typed programming languages (like Fortran or C), since the data-type of the result is the same as the inputs, but the former is more inline with our expectations.
    
</div>

In [6]:
1/2

0.5

To get an integer result, we can use the // operator.

In [7]:
1//2

0

Python is a _dynamically-typed language_&mdash;this means that we do not need to declare the datatype of a variable before initializing it.  

Here we'll create a variable (think of it as a descriptive label that can refer to some piece of data).  The `=` operator assigns a value to a variable.  

In [8]:
a = 1
b = 2

Functions operate on variables and return a result.  Here, `print()` will output to the screen.

In [9]:
print(a+b)

3


In [10]:
print(a*b)

2


Note that variable names are case sensitive, so a and A are different

In [11]:
A = 2048

In [12]:
print(a, A)

1 2048


Here we initialize 3 variables all to `0`, but these are still distinct variables, so we can change one without affecting the others.

In [13]:
x = y = z = 0

In [14]:
print(x, y, z)

0 0 0


In [15]:
z = 1

In [16]:
print(x, y, z)

0 0 1


Python has some built in help (and Jupyter/ipython has even more)

In [17]:
help(x)

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil_

In [18]:
x?

Another function, `type()` returns the data type of a variable

In [19]:
print(type(x))

<class 'int'>


Note in languages like Fortran and C, you specify the amount of memory an integer can take (usually 2 or 4 bytes).  This puts a restriction on the largest size integer that can be represented.  Python will adapt the size of the integer so you don't *overflow*

In [20]:
a = 12345678901234567890123456789012345123456789012345678901234567890
print(a)
print(a.bit_length())
print(type(a))

12345678901234567890123456789012345123456789012345678901234567890
213
<class 'int'>


## floating point

when operating with both floating point and integers, the result is promoted to a float.  This is true of both python 2.x and 3.x

In [21]:
1./2

0.5

but note the special integer division operator

In [22]:
1.//2

0.0

It is important to understand that since there are infinitely many real numbers between any two bounds, on a computer we have to approximate this by a finite number.  There is an IEEE standard for floating point that pretty much all languages and processors follow.  

The means two things

* not every real number will have an exact representation in floating point
* there is a finite precision to numbers -- below this we lose track of differences (this is usually called *roundoff* error)

This paper: [What every computer scientist should know about floating-point arithmetic](https://dl.acm.org/doi/10.1145/103162.103163) is a great reference on understanding how a computer stores numbers.

Consider the following expression, for example:

In [23]:
0.3/0.1 - 3

-4.440892098500626e-16

Here's another example: The number 0.1 cannot be exactly represented on a computer.  In our print, we use a format specifier (the stuff inside of the {}) to ask for more precision to be shown:

In [24]:
a = 0.1
print("{:30.20}".format(a))

        0.10000000000000000555


we can ask python to report the limits on floating point

In [25]:
import sys
print(sys.float_info)

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)


Note that this says that we can only store numbers between 2.2250738585072014e-308 and 1.7976931348623157e+308

We also see that the precision is 2.220446049250313e-16 (this is commonly called _machine epsilon_).  To see this, consider adding a small number to 1.0.  We'll use the equality operator (`==`) to test if two numbers are equal:

<div class="alert alert-block alert-info">
<h3><span class="fa fa-flash"></span> Quick Exercise:</h3>

1. Define two variables, $a = 1$, and $e = 10^{-16}$.

2. Now define a third variable, `b = a + e`

3. We can use the python `==` operator to test for equality.  
    
What do you expect `b == a` to return?
    
Run it an see if it agrees with your guess.
</div>

In [35]:
a = 1
e = 1e-16
b = a + e
b == a
print(b == a)

True


In [34]:
1.e-6

1e-06

## modules

The core python language is extended by a standard library that provides additional functionality.  These added pieces are in the form of modules that we can _import_ into our python session (or program).

The `math` module provides functions that do the basic mathematical operations as well as provide constants (note there is a separate `cmath` module for complex numbers).

In python, you `import` a module.  The functions are then defined in a separate _namespace_&mdash;this is a separate region that defines names and variables, etc.  A variable in one namespace can have the same name as a variable in a different namespace, and they don't clash.  You use the "`.`" operator to access a member of a namespace.

By default, when you type stuff into the python interpreter or here in the Jupyter notebook, or in a script, it is in its own default namespace, and you don't need to prefix any of the variables with a namespace indicator.

In [36]:
import math

`math` provides the value of pi

In [37]:
print(math.pi)

3.141592653589793


This is distinct from any variable `pi` we might define here

In [38]:
pi = 3

In [39]:
print(pi, math.pi)

3 3.141592653589793


Note here that `pi` and `math.pi` are distinct from one another&mdash;they are in different namespaces.

### floating point operations

The same operators, `+`, `-`, `*`, `/` work are usual for floating point numbers.  To raise an number to a power, we use the `**` operator (this is the same as Fortran)

In [40]:
R = 2.0

In [42]:
print(math.pi*R**2)

12.566370614359172


operator precedence follows that of most languages.  See

https://docs.python.org/3/reference/expressions.html#operator-precedence
    
in order of precedence:
* quantites in `()`
* slicing, calls, subscripts
* exponentiation (`**`)
* `+x`, `-x`, `~x`
* `*`, `@`, `/`, `//`, `%`
* `+`, `-`

(after this are bitwise operations and comparisons)

Parantheses can be used to override the precedence.

<div class="alert alert-block alert-info">

<h3><span class="fa fa-flash"></span> Quick Exercise:</h3>

Consider the following expressions.  Using the ideas of precedence, think about what value will result, then try it out in the cell below to see if you were right.

  * `1 + 3*2**2`
  * `1 + (3*2)**2`
  * `2**3**2`

</div>

In [47]:
2**(3**2)

512

The `math` module provides a lot of the standard math functions we might want to use.

For the trig functions, the expectation is that the argument to the function is in radians&mdash;you can use `math.radians()` to convert from degrees to radians, ex:

In [48]:
print(math.cos(math.radians(45)))

0.7071067811865476


Notice that in that statement we are feeding the output of one function (`math.radians()`) into a second function, `math.cos()`

When in doubt, as for help to discover all of the things a module provides:

In [50]:
help(math)

Help on module math:

NAME
    math

MODULE REFERENCE
    https://docs.python.org/3.8/library/math
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module provides access to the mathematical functions
    defined by the C standard.

FUNCTIONS
    acos(x, /)
        Return the arc cosine (measured in radians) of x.
    
    acosh(x, /)
        Return the inverse hyperbolic cosine of x.
    
    asin(x, /)
        Return the arc sine (measured in radians) of x.
    
    asinh(x, /)
        Return the inverse hyperbolic sine of x.
    
    atan(x, /)
        Return the arc tangent (measured in radians) of x.
    
    atan2(y, x, /)
        Return the arc tangent (measured in radians) of y/x.
    

## complex numbers

python uses '`j`' to denote the imaginary unit

In [51]:
print(1.0 + 2j)

(1+2j)


In [52]:
a = 1j
b = 3.0 + 2.0j
print(a + b)
print(a*b)

(3+3j)
(-2+3j)


we can use `abs()` to get the magnitude and separately get the real or imaginary parts 

In [53]:
print(abs(b))
print(a.real)
print(a.imag)

3.605551275463989
0.0
1.0


## strings

python doesn't care if you use single or double quotes for strings:

In [54]:
a = "this is my string"
b = 'another string'

In [55]:
print(a)
print(b)

this is my string
another string


Many of the usual mathematical operators are defined for strings as well.  For example to concatenate or duplicate:

In [56]:
print(a+b)

this is my stringanother string


In [57]:
print(a + ". " + b)

this is my string. another string


In [58]:
print(a*2)

this is my stringthis is my string


There are several escape codes that are interpreted in strings.  These start with a backwards-slash, `\`.  E.g., you can use `\n` for new line

In [60]:
a = a + "\n\n"
print(a)

this is my string





<div class="alert alert-block alert-info">
    
<h3><span class="fa fa-flash"></span> Quick Exercise:</h3>

The `input()` function can be used to ask the user for input.

  * Use `help(input)` to see how it works.  
  * Write code to ask for input and store the result in a variable.  `input()` will return a string.

  * Use the `float()` function to convert a number entered as input to a floating point variable.  
  * Check to see if the conversion worked using the `type()` function.
</div>

In [63]:
a = input("enter a number:")

enter a number:5.0


In [66]:
b = float(a)
type(b)

float

In [67]:
b

5.0

""" can enclose multiline strings.  This is useful for docstrings at the start of functions (more on that later...)

In [68]:
c = """
Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore 
eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt 
in culpa qui officia deserunt mollit anim id est laborum."""

In [69]:
print(c)


Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore 
eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt 
in culpa qui officia deserunt mollit anim id est laborum.


a raw string does not replace escape sequences (like \n).  Just put a `r` before the first quote:

In [70]:
d = r"this is a raw string\n"
print(d)

this is a raw string\n


slicing is used to access a portion of a string.

slicing a string can seem a bit counterintuitive if you are coming from Fortran.  The trick is to think of the index as representing the left edge of a character in the string.  When we do arrays later, the same will apply.

Also note that python (like C) uses 0-based indexing

Negative indices count from the right.

In [71]:
a = "this is my string"
print(a)
print(a[5:7])
print(a[0])
print(d)
print(d[-2])

this is my string
is
t
this is a raw string\n
\


<div class="alert alert-block alert-info">
    
<h3><span class="fa fa-flash"></span> Quick Exercise:</h3>

Strings have a lot of _methods_ (functions that know how to work with a particular datatype, in this case strings).  A useful method is `.find()`.  For a string `a`,
`a.find(s)` will return the index of the first occurrence of `s`.

For our string `c` above, find the first `.` (identifying the first full sentence), and print out just the first sentence in `c` using this result

</div>

In [73]:
idx = c.find(".")
print(idx)

125


In [76]:
c[:idx+1]

'\nLorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor \nincididunt ut labore et dolore magna aliqua.'

There are also a number of methods and functions that work with strings.  Here are some examples:

In [ ]:
print(a.replace("this", "that"))
print(len(a))
print(a.strip())    # Also notice that strip removes the \n
print(a.strip()[-1])

Note that our original string, `a`, has not changed.  In python, strings are *immutable*.  Operations on strings return a new string.

In [ ]:
print(a)

In [ ]:
print(type(a))

As usual, ask for help to learn more:

In [ ]:
help(str)

We can format strings when we are printing to insert quantities in particular places in the string.  A `{}` serves as a placeholder for a quantity and is replaced using the `.format()` method:

In [ ]:
a = 1
b = 2.0
c = "test"
print("a = {}; b = {}; c = {}".format(a, b, c))